initrd/bin/oem-factory-reset

```diff
@@ -437,343 +833,444 @@ usb_security_token_capabilities_check
-	if [ -n "$luks_new_Disk_Recovery_Key_passphrase_desired" -o -n "$luks_new_Disk_Recovery_Key_passphrase" ]; then
-		CUSTOM_PASS_AFFECTED_COMPONENTS="LUKS Disk Recovery Key passphrase"
-	fi
-	if [ "$CONFIG_TPM" = "y" ]; then
-		CUSTOM_PASS_AFFECTED_COMPONENTS="$CUSTOM_PASS_AFFECTED_COMPONENTS
-TPM Ownership password"
-	fi
-	CUSTOM_PASS_AFFECTED_COMPONENTS="$CUSTOM_PASS_AFFECTED_COMPONENTS
-GPG Admin PIN
-GPG User PIN"
-
-	# Inform user of security components affected for the following prompts
-	echo -e "The following security components will be provisioned with defaults or chosen PINs/passwords:
+    #Give general guidance to user on how to answer prompts
+    echo -e "\n **** Factory Reset / Re-Ownership Questionnaire ****\n\n"
+    echo -e "The following questionnaire will help you configure the security components of your system.\n"
+    echo -e "Each prompt requires a single letter answer: eg. (Y/n).\n"
+    echo -e "If you don't know what to answer, pressing Enter will select the default answer for that prompt: eg. Y, above.\n"
+
+    # Re-ownership of encrypted disk key, content and passphrase
+    echo -e -n "\n\nWould you like to change the current LUKS Disk Recovery Key passphrase?\n (Highly recommended if you didn't install the Operating System yourself, so that past provisioned passphrase would not permit to access content.\n  Note that without re-encrypting disk, a backed up header could be restored to access encrypted content with old passphrase) [y/N]: "
+    read -n 1 prompt_output
+    echo
+    if [ "$prompt_output" == "y" \
+        -o "$prompt_output" == "Y" ]; then
+        luks_new_Disk_Recovery_Key_passphrase_desired=1
+        echo -e "\n"
+    fi
+
+    echo -e -n "Would you like to re-encrypt LUKS encrypted container and generate new Disk Recovery key?\n (Highly recommended if you didn't install the operating system yourself: this would prevent any LUKS backed up header to be restored to access encrypted data) [y/N]: "
+    read -n 1 prompt_output
+    echo
+    if [ "$prompt_output" == "y" \
+        -o "$prompt_output" == "Y" ]; then
+        test_luks_current_disk_recovery_key_passphrase
+        luks_new_Disk_Recovery_Key_desired=1
+        echo -e "\n"
+    fi
+
+    #Prompt to ask if user wants to generate GPG key material in memory or on smartcard
+    echo -e -n "Would you like to format an encrypted USB Thumb drive to store GPG key material generated in memory?\n (Required to enable GPG authentication) [y/N]: "
+    read -n 1 prompt_output
+    echo
+    if [ "$prompt_output" == "y" \
+        -o "$prompt_output" == "Y" ] \
+        ; then
+        GPG_GEN_KEY_IN_MEMORY=1
+        echo " ++++ Master key and subkeys will be generated in memory, backed up to dedicated LUKS container +++"
+        echo -e -n "Would you like in-memory generated subkeys to be copied to USB Security Dongle' smartcard?\n (Highly recommended so the smartcard is used on daily basis and backup is kept safe, but not required) [Y/n]: "
```

Comment:
OK, so the intent here is to allow use of just the encrypted flash drive with no USB token.

Please add the missing `s` in `Dongle'` :wink: 

I do want to point out that we are up to a max of 10x Y/N prompts (depending a little on config and selections).

I am good with this prompt here as it'd be pretty hard to offer this configuration later without a lot more development, so for a first version of this feature it's OK.  In the long run I think we should try to find ways to reduce the amount of prompts here though :thinking:  In my view the OEM reset serves as the fastest way to get from an unconfigured or partially configured state to a working state, so things that could be configured later/optionally are better served with configuration settings.  That's more flexible too as it allows you to change the configuration later, which OEM reset does not serve well.

Reply:
> Please add the missing s in Dongle' 😉

fixed under [1ce158b](https://github.com/linuxboot/heads/pull/1515/commits/1ce158b1c1f0515fab5d0a41d7eae88f03552e06)

Reply:
> I'm not sure how to tackle this correctly. I think in case `CONFIG_OEMRESET_OFFER_DEFAULTS` is not defined, maybe the questionnaire's first prompt should be that same question asked to all as a start:

That's what CONFIG_OEMRESET_OFFER_DEFAULTS enables - the offer to select all defaults and skip all the questions.  If you enable it, then the user says 'n', all the questions are asked normally.

If you want to offer defaults first on all boards, you could delete CONFIG_OEM_OFFER_DEFAULTS and always ask that first question.  All the Purism boards already have this config enabled so it wouldn't change anything for me :grin:

Reply:
@tlaurion Left this unresolved so you could see the answers to your question, but like I said I'm OK with the current state of this PR.  Opened #1521 for discussion on the future of OEM reset.

The CONFIG_OEMRESET_OFFER_DEFAULTS change would be simple but does not have to be done in this PR unless you want to (doesn't affect me since all my boards enable it already).